apollo-federation/src/sources/connect/fetch_dependency_graph/mod.rs

```diff
@@ -147,28 +372,75 @@ pub(crate) struct PathField {
     selections: PathSelections,
 }
 
+/// A path to a specific selection
+///
+/// This enum encompasses a set of directions for reaching a target property
+/// within a JSON selection.
 #[derive(Debug, Clone)]
 pub(crate) enum PathSelections {
+    /// Set of selections assuming a starting point of the root.
     Selections {
+        /// Property path from the head
+        ///
+        /// This is a list of properties to traverse starting from the root (or head)
+        /// of the corresponding selection. These should be simple paths that can be
+        /// chained together.
         head_property_path: Vec<Key>,
+
+        /// Named selections from the root reachable through [head_property_path]
+        ///
+        /// Each member in this list is of the form ([Name], [Keys](Vec<Key>)) and is
+        /// a chain of named selections to apply iteratively to reach the leaf of our selection.
+        ///
+        /// Note: [Name] here can refer to aliased fields as well.
         named_selections: Vec<(Name, Vec<Key>)>,
+
+        /// The (optional) final selection for this chain.
+        ///
+        /// This selection is assumed to be from the context of the node accessable from
+        /// the chain of [head_property_path] followed by the chain of [named_selections].
+        ///
+        /// A value of `None` here means to stop traversal at the current selection, while any
+        /// other value signals that there might be further sections to traverse.
         tail_selection: Option<(Name, PathTailSelection)>,
     },
+
+    /// The full selection from a (potentially) different root
     CustomScalarRoot {
+        /// The full selection
         selection: JSONSelection,
     },
 }
 
+/// A path to a specific selection, not from the root.
+///
+/// This enum describes different ways to perform a final selection of a path
+/// from the context of any selection in the tree.
 #[derive(Debug, Clone)]
 pub(crate) enum PathTailSelection {
+    /// Simple selection using a chain of keys.
     Selection {
+        /// The chain of [Key]s to traverse
         property_path: Vec<Key>,
     },
-    CustomScalarPathSelection {
-        path_selection: PathSelection,
-    },
+
+    /// Custom selection using a [PathSelection]
```

Comment:
I'm not sure if "custom selection" is the right phrase here, it's more of a selection that returns a GraphQL custom scalar (similarly below).